### PR TITLE
Single queue

### PR DESCRIPTION
**Pull request by krulis-martin**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -15,7 +15,7 @@ include_directories(AFTER, vendor/spdlog/include)
 find_package(CURL REQUIRED)
 include_directories(${CURL_INCLUDE_DIRS})
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -Wall")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17 -Wall")
```
Comment (Teyras) [inline CMakeLists.txt:18]: Bleeding edge, aren't we?

File: README.md [hunk 2]
```diff
@@ -144,6 +144,7 @@ The default location for broker configuration file is
 	  `err`, `warn`, `notice`, `info` and `debug`
 	- _max-size_ -- maximal size of log file before rotating
 	- _rotations_ -- number of rotation kept
+- _queue_manager_ -- selection of the queue manager implementation responsible for assigning jobs to workers. Currently only `single` (the default) and `multi` queue managers are in production version. Single-queue manager has one queue and dispatches jobs on demand as workers become available. Multi-queue manager has a queue for every worker, jobs are assigned immediately and cannot be re-assigned unless failure occurs. I.e., `single` provides better load balancing, `mutli` has lower dispatching overhead.
```
Comment (Teyras) [inline README.md:147]: Tpyo

File: src/broker_core.cpp [hunk 3]
```diff
@@ -127,7 +128,17 @@ void broker_core::broker_init()
 	logger_->info("Initializing broker connection...");
 	workers_ = std::make_shared<worker_registry>();
 	context_ = std::make_shared<zmq::context_t>(1);
-	queue_ = std::make_shared<multi_queue_manager>();
+
+	// Yes, this may be done better, but it will do for now.
```
Comment (Teyras) [inline src/broker_core.cpp:132]: If I were pedantic, I would ask you to elaborate on what's the better way. But I'm not, so keep your secrets if you want.

File: src/queuing/queue_manager_interface.h [hunk 4]
```diff
@@ -80,7 +80,7 @@ class queue_manager_interface
 	virtual request_ptr worker_finished(worker_ptr worker) = 0;
 
 	/**
-	 * Mark the current request of a worker as cancelled and do not assign it another request.
+	 * Mark the current request of a worker as cancelled and do not (re)assign it to another worker.
```
Comment (Teyras) [inline src/queuing/queue_manager_interface.h:83]: While you're at it, why not make it clear that reassigning the job is a responsibility of the caller?

File: src/queuing/single_queue_manager.h [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    std::chrono::milliseconds arrived_at;
+};
+
+struct fcfs_job_comparator {
+    bool compare(const request_entry &a, const request_entry &b, worker_ptr worker) const
+    {
+        return a.arrived_at < b.arrived_at;
+    }
+
+    void drop_caches() {}
```
Comment (Teyras) [inline src/queuing/single_queue_manager.h:23]: I believe we can safely remove this - it was there to speed up the simulation of another comparator (OAGM, if I recall correctly), which wasn't that great anyways.

File: src/queuing/single_queue_manager.h [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        }
+        worker_jobs_.erase(worker);
+        workers_.erase(std::remove(workers_.begin(), workers_.end(), worker), workers_.end());
+
+        // filter jobs and remove those wich are no longer process-able (after worker removal)
+        for (auto &&job : jobs_) {
+            if (!is_request_assignable(job.request)) {
+                // the job cannot be accomodated anymore...
+                result->push_back(job.request);
+                job.request = nullptr; // mark the job for removal
```
Comment (Teyras) [inline src/queuing/single_queue_manager.h:129]: That's... unusual. In such short code though... I guess it's alright

**Review by janbuchar — DISMISSED** [review]
Seems legit!

File: CMakeLists.txt (hunk 1 above)
Comment (krulis-martin) [reply CMakeLists.txt:18]: Not really, but I suppose that GCC 8.x (which is the default in CentOS-like systems now) should be fine with this.

File: src/broker_core.cpp (hunk 3 above)
Comment (krulis-martin) [reply src/broker_core.cpp:132]: You can create a static map<string, obj>, where the obj is either factory, prototype, or singleton. This map can be then filled by static constructors (so the "registration" for translation is done along with the declaration and will not be forgotten when extensions are implemented). However, for 2 or 3 manager classes, it does not worth implementing.

File: src/queuing/queue_manager_interface.h (hunk 4 above)
Comment (krulis-martin) [reply src/queuing/queue_manager_interface.h:83]: That is two lines below :)

File: src/broker_core.cpp (hunk 3 above)
Comment (Teyras) [reply src/broker_core.cpp:132]: heh, that sounds like so much of an overkill I'd even remove the comment

File: src/queuing/queue_manager_interface.h (hunk 4 above)
Comment (Teyras) [reply src/queuing/queue_manager_interface.h:83]: Indeed it is, my bad

File: src/queuing/single_queue_manager.h (hunk 6 above)
Comment (krulis-martin) [reply src/queuing/single_queue_manager.h:129]: What is unusual? The marking for deletion? If you have a better way, I would be happy to entertain it, but I am quite sure that deleting items from a container in a for-each loop (over that container) is not something you want to do...

File: src/broker_core.cpp (hunk 3 above)
Comment (krulis-martin) [reply src/broker_core.cpp:132]: Well, I have a personal grudge against extensive if-else branching, so I shall keep it :)

File: src/queuing/single_queue_manager.h (hunk 6 above)
Comment (Teyras) [reply src/queuing/single_queue_manager.h:129]: If I went with the "mark for deletion" route, I'd probably make a separate collection of stuff to remove. As I said, the current implementation works just fine and it's unlikely the method will ever grow enough for it to become unmaintainable.

An alternative idea I have is to use `stable_partition` to push the unsatisfiable request to the edge of the queue and then snip them off.

Comment (krulis-martin) [reply src/queuing/single_queue_manager.h:129]: that's what remove_if does ... I guess I have become too biased towards functional JS-like solutions ;)

Comment (Teyras) [reply src/queuing/single_queue_manager.h:125]: well, you can call mutating the queue items in place many names, but functional isn't one of them :slightly_smiling_face:

Comment (krulis-martin) [reply src/queuing/single_queue_manager.h:125]: Yep, I mutated the objects inside, but the collection was intact until filtered by erase-remove, which is functional (that's what I meant).

**Review by janbuchar — APPROVED** [review]

**Pull request merged**